sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -0,0 +1,204 @@
+      toType: IntegralType,
+      value: Any): Expression = {
+
+    val fromType = fromExp.dataType
+    val (min, max) = getRange(fromType)
+    val (minInToType, maxInToType) = {
+      (Cast(Literal(min), toType).eval(), Cast(Literal(max), toType).eval())
+    }
+    val ordering = toType.ordering.asInstanceOf[Ordering[Any]]
+    val minCmp = ordering.compare(value, minInToType)
+    val maxCmp = ordering.compare(value, maxInToType)
+    val lit = Cast(Literal(value), fromType)
+
+    (minCmp.signum, maxCmp.signum, exp) match {
+      case (_, 1, EqualTo(_, _) | GreaterThan(_, _) | GreaterThanOrEqual(_, _)) =>
+        fromExp.falseIfNotNull
+      case (_, 1, LessThan(_, _) | LessThanOrEqual(_, _)) =>
+        fromExp.trueIfNotNull
+      // make sure the expression is evaluated if it is non-deterministic
+      case (_, 1, EqualNullSafe(_, _)) if exp.deterministic =>
+        FalseLiteral
+      case (_, 1, _) => exp // impossible but safe guard, same below
+
+      case (_, 0, GreaterThan(_, _)) =>
+        fromExp.falseIfNotNull
+      case (_, 0, LessThanOrEqual(_, _)) =>
+        fromExp.trueIfNotNull
+      case (_, 0, LessThan(_, _)) =>
+        Not(EqualTo(fromExp, Literal(max, fromType)))
+      case (_, 0, GreaterThanOrEqual(_, _) | EqualTo(_, _) | EqualNullSafe(_, _)) =>
+        EqualTo(fromExp, Literal(max, fromType))
+      case (_, 0, _) => exp
+
+      case (-1, _, GreaterThan(_, _) | GreaterThanOrEqual(_, _)) =>
+        fromExp.trueIfNotNull
+      case (-1, _, LessThan(_, _) | LessThanOrEqual(_, _) | EqualTo(_, _)) =>
+        fromExp.falseIfNotNull
+      // make sure the expression is evaluated if it is non-deterministic
+      case (-1, _, EqualNullSafe(_, _)) if exp.deterministic =>
+        FalseLiteral
+      case (-1, _, _) => exp
+
+      case (0, _, LessThan(_, _)) =>
+        fromExp.falseIfNotNull
+      case (0, _, GreaterThanOrEqual(_, _)) =>
+        fromExp.trueIfNotNull
+      case (0, _, GreaterThan(_, _)) =>
+        Not(EqualTo(fromExp, Literal(min, fromType)))
+      case (0, _, LessThanOrEqual(_, _) | EqualTo(_, _) | EqualNullSafe(_, _)) =>
+        EqualTo(fromExp, Literal(min, fromType))
+      case (0, _, _) => exp
+
+      case (_, _, GreaterThan(_, _)) => GreaterThan(fromExp, lit)
+      case (_, _, GreaterThanOrEqual(_, _)) => GreaterThanOrEqual(fromExp, lit)
+      case (_, _, EqualTo(_, _)) => EqualTo(fromExp, lit)
+      case (_, _, EqualNullSafe(_, _)) => EqualNullSafe(fromExp, lit)
+      case (_, _, LessThan(_, _)) => LessThan(fromExp, lit)
+      case (_, _, LessThanOrEqual(_, _)) => LessThanOrEqual(fromExp, lit)
+      case (_, _, _) => exp
+    }
+  }
+
+  /**
+   * Check if the input `fromExp` can be safely cast to `toType` without any loss of precision,
+   * i.e., the conversion is injective. Note this only handles the case when both sides are of
+   * integral type.
+   */
+  private def canImplicitlyCast(fromExp: Expression, toType: DataType): Boolean = {
```

Comment:
Nit, in general, we had better narrow down the parameter from `fromExp` to `fromType`. But, it seems that there is a follow-up plan to extend this optimizer. This looks fine in that case.

Reply:
Yes this method is evolving and will be enhanced in the follow-ups.